### ISSUE TITLE
some questions about the model size

### ISSUE THREAD

**Issue by youngboy52** [ZiweiWangTHU/BiDet#13]
Hi, could you tell me the size of your ssd or faster rcnn model? I found that my own trained faster rcnn model takes 142.14MB space! It is still too large.

**Comment by Wuziyi616** [comment]
Thank you for your attention to our work! Yes, the trained models in this code are large because we don't perform **binarization** in our code. PyTorch doesn't support computation with *binary* weight so far. If you want to get a smaller model from a model trained using this repo, I recommend you to refer to some BNN frameworks (e.g. [daBNN](https://github.com/JDAI-CV/dabnn)).

Also, I need to point out that, we don't implement the bi-real re-training schedule in this repo. So in the Bi-Real Net paper, the authors first train the model with *real-valued* weight, then re-train the model with learning_rate=0 to let the BN absorb the magnitude of the weight. Thus you can't directly binarize the weight of the models trained using this repo. You can refer to [bi-real](https://github.com/liuzechun/Bi-Real-net) for the re-training implementation.

**Comment by youngboy52** [comment]
Thank you for your reply, I read the BiDet paper, and found that the experimental results reported in your paper include the **BiDet**, **Bi-Real-Net** and **BiDet (SC)**. According to this repo, the trained **faster rcnn** model based on resent-18 can obtained mAP 57.42 in PASCAL VOC dataset, which is close to the result (Table 1) of Bi-Real-Net in your paper. So could you tell me some details about the implementations of **BiDet** and **BiDet(SC)**? are they also trained in real value? Looking forward to your reply!
![image](https://user-images.githubusercontent.com/15629424/93190662-3542e700-f776-11ea-945a-d2e609359c45.png)

**Comment by Wuziyi616** [comment]
Hi! So basically BiDet means using Xnor-Net as the **detector architecture** (quantization method, network connection, etc.) plus the *Sparse Object Prior* and *IB* training loss. And BiDet (SC) means using Bi-Real-Net as the detector architecture plus our proposed losses (SC stands for shortcut, which is adopted in Bi-Real-Net). In this repo, we only implement BiDet (SC) because it achieves better mAP than BiDet.

So, if you see something like "reg loss: xxx, prior loss: xxx", then this is BiDet (SC) training. I think the reason you get worse mAP than reported in the paper is because of the training schedule, e.g. learning rate decay and batch size. You can refer to [this issue](https://github.com/ZiweiWangTHU/BiDet/issues/8#issuecomment-664217441) for more details about training binary detectors.

Just to make sure one more thing, all the "Xnor-Net", "Bi-Real-Net", "BiDet", "BiDet (SC)" are trained using binary values.

**Comment by youngboy52** [comment]
Hi! sorry to bother you again, as you said, these models are trained using binary values. In this repo, can I train BiDet (SC) using binary values？According to the code, this repo uses nn.Conv2d instead of your defined BinarizeConv2d in class BiDetResnet, however RPN_Conv uses the BinarizeConv2d. It confuses me that whether BiDet (SC) are trained using binary values in this repo. If I want to obtain a binarized network, do I need to transfer the trained BiDet (SC) model to binarized one using daBNN?

**Comment by Wuziyi616** [comment]
Oh, I understand what you mean. It is a common practice in BNN that, we keep the **first** and **last** layer of the network full-precision (see Xnor-Net, Bi-Real-Net, etc.). That's why I use nn.Conv2d [here](https://github.com/ZiweiWangTHU/BiDet/blob/master/faster_rcnn/lib/model/faster_rcnn/bidet_resnet.py#L88). Besides, Bi-Real-Net also uses full-precision Conv in the **downsample residual connection** of ResNet, and that's why we also use nn.Conv2d [here](https://github.com/ZiweiWangTHU/BiDet/blob/master/faster_rcnn/lib/model/faster_rcnn/bidet_resnet.py#L115). But in other layers of the network, we use BinaryConv, for example [here](https://github.com/ZiweiWangTHU/BiDet/blob/master/faster_rcnn/lib/model/faster_rcnn/bidet_resnet.py#L149), you can see that we feed the BinBasicBlock which consists of two BinaryConv into the function. Also as you point out, we use BinaryConv in the RPN.

One more thing, I'm not familiar with daBNN. But I think if you want to get a binarized network, whose weight is -1 and +1 in most of the layers, you need to re-train the network by only updating the BN to absorb the magnitude caused by binarization. You can refer to the Bi-Real-Net [paper](https://openaccess.thecvf.com/content_ECCV_2018/papers/zechun_liu_Bi-Real_Net_Enhancing_ECCV_2018_paper.pdf), Section 4.1, "Training" part for more details.
